lib/turbolinks/x_domain_blocker.rb

```diff
@@ -4,18 +4,18 @@ module Turbolinks
   # - The request is being redirected to a different domain
   module XDomainBlocker
     private
-      def same_origin?(a, b)
-        a = URI.parse URI.escape(a)
-        b = URI.parse URI.escape(b)
-        [a.scheme, a.host, a.port] == [b.scheme, b.host, b.port]
-      end
-
       def abort_xdomain_redirect
-        to_uri = response.headers['Location'] || ""
-        current = request.headers['X-XHR-Referer'] || ""
-        unless to_uri.blank? || current.blank? || same_origin?(current, to_uri)
+        if _turbolinks_xdomain_redirect?
           self.status = 403
         end
       end
+
+      def _turbolinks_xdomain_redirect?
+        from = request.headers['X-XHR-Referer']
+        to = response.headers['Location']
+        from.present? && to.present? && URI(from).select(:scheme, :host, :port) != URI(to).select(:scheme, :host, :port)
+      rescue URI::InvalidURIError
+        false
+      end
```

Comment:
I don't think we need to `URI.escape` anymore since we're only looking at `scheme`, `host` and `port`. But maybe I'm missing something?

`URI.escape` doesn't guarantee a valid URI string, e.g. `URI.escape(':') => ':'`.